src/collapse/dtrace.rs

```diff
@@ -221,83 +231,108 @@ impl Folder {
-                            loop {
-                                if rx_input.recv().unwrap().is_none() {
-                                    break;
+                    // Loop until the main thread signals there is no more data
+                    // or another worker signals they have errored...
+                    loop {
+                        channel::select! {
+                            recv(rx_input) -> input => {
+                                let data = match input.unwrap() {
+                                    Some(data) => data,
+                                    None => return,
+                                };
+                                if let Err(e) = folder.collapse_single_threaded(&data[..]) {
+                                    // If we produce an error...
+
+                                    // Signal all the other threads; so they can stop doing needless work.
+                                    // (note: if this fails, it means another thread has already sent
+                                    // stop signals to all the other threads; so we don't have to.)
+                                    for _ in 0..(folder.opt.nthreads - 1) {
+                                        let _ = tx_stop.try_send(());
+                                    }
+
+                                    // Send the error back to the main thread for propagation.
+                                    // (note: if this fails it means another thread has already sent an error
+                                    // to the main thread; so we don't have to.)
+                                    let _ = tx_output.try_send(Err(e));
+
+                                    // Return.
+                                    return;
                                 }
-                            }
+                            },
+                            recv(rx_stop) -> _ => return,
                         }
                     }
                 });
                 handles.push(handle);
             }
 
-            // State for the loop...
+            drop(rx_input);
+
             let buf_capacity =
                 usize::next_power_of_two(collapse::NBYTES_PER_STACK_GUESS * self.nstacks_per_job);
             let mut buf = Vec::with_capacity(buf_capacity);
             let (mut index, mut nstacks) = (0, 0);
 
-            // The loop...
+            // Loop through the input data in order to chunk it up and send it off to the worker threads...
             loop {
-                // First, read some data into the `utils::Buffer`.
+                // First, read some data into the buffer.
                 let n = reader.read_until(b'\n', &mut buf)?;
 
                 // If we're at the end of the data...
                 if n == 0 {
                     // Send the last slice and exit the loop.
-                    tx_input.send(Some(buf)).unwrap();
+                    let _ = tx_input.send(Some(buf));
```

Comment:
I'd like to see a comment as to why we're ignoring the error here, like for the things above.